Don't write DNS resolution errors on Test-Connection -Quiet

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

We don't need to write a DNS resolution error if the user uses `-Quiet`; it's the same as a `False` result in that case and can be treated the same. 

Also, currently we're not able to respond to StopProcessing() between pings due to the `Dns.GetHostEntry()` method blocking. Fix is to implement a `GetHostEntryWithCancel()` method that factors in a cancellation token.

## PR Context

Resolves #12147 
Resolves #11160  

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:bug: Don't write hostname resolve error on -Quiet

vexx32 committed:
:sparkles: Add cancellation for DNS lookups

+ Add GetHostEntryWithCancel method and _cancel member.

+ Add cancellation for DNS lookups.

PoshChan commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -58,6 +58,8 @@ public class TestConnectionCommand : PSCmdlet, IDisposable
 
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
Test-Connection.Ping.Ping fake host
```powershell
Expected an exception, with FullyQualifiedErrorId 'TestConnectionException,Microsoft.PowerShell.Commands.TestConnectionCommand' to be thrown, but no exception was thrown.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 65
65:                 Should -Throw -ErrorId "TestConnectionException,Microsoft.PowerShell.Commands.TestConnectionCommand"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
Test-Connection.Ping.Ping fake host
```powershell
Expected an exception, with FullyQualifiedErrorId 'TestConnectionException,Microsoft.PowerShell.Commands.TestConnectionCommand' to be thrown, but no exception was thrown.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 65
65:                 Should -Throw -ErrorId "TestConnectionException,Microsoft.PowerShell.Commands.TestConnectionCommand"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
Test-Connection.Ping.returns false without errors for an unresolvable address when using -Quiet
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 65
65:             $result | Should -BeFalse
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.Ping.returns false without errors for an unresolvable address when using -Quiet
```powershell
Expected $false, but got $true.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 65
65:             $result | Should -BeFalse
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
Test-Connection.Ping.returns false without errors for an unresolvable address when using -Quiet
```powershell
Expected $false, but got $true.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 65
65:             $result | Should -BeFalse
```

vexx32 committed:
:white_check_mark: Update tests

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -59,18 +59,19 @@ Describe "Test-Connection" -tags "CI" {
-        It "Ping fake host" {
+        It 'does not emit errors for an unresolvable address when using -Quiet' {
+            Test-Connection -Quiet -TargetName "fakeHost" -Count 1 | Should -BeFalse
+        }
```
The test name is seem not right.

vexx32 replied:
I'll give it another pass, I missed a couple things here, it seeems.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -679,16 +682,26 @@ private void ProcessPing(string targetNameOrAddress)
+                    if (ex is PipelineStoppedException p)
+                    {
+                        throw p;
+                    }
```
Please move this in new `catch` block.
```c#
catch (PipelineStoppedException ex)
{
    throw ex;
}
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -732,6 +745,23 @@ private void ProcessPing(string targetNameOrAddress)
 
+        private IPHostEntry GetHostEntryWithCancel(string targetNameOrAddress)
+        {
+            _dnsLookupCancel = new CancellationTokenSource();
```
Is the method called in all code paths? In the case please move the initializer in the field definition.

vexx32 replied:
I'm... not sure if it is, actually. But thinking about it, I'd be surprised if it was not. I'll move it. 🙂

vexx32 committed:
:white_check_mark: Update test

vexx32 committed:
:ok_hand: Address review comments

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -58,6 +58,8 @@ public class TestConnectionCommand : PSCmdlet, IDisposable
 
         private static byte[]? s_DefaultSendBuffer;
 
+        private CancellationTokenSource? _dnsLookupCancel = new CancellationTokenSource();
```
No need to make it nullable.

vexx32 replied:
```suggestion
        private CancellationTokenSource _dnsLookupCancel = new CancellationTokenSource();
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -59,18 +59,20 @@ Describe "Test-Connection" -tags "CI" {
-        It "Ping fake host" {
+        It 'returns false without errors for an unresolvable address when using -Quiet' {
+            { Test-Connectiong -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" } | Should -Not -Throw
+            Test-Connection -Quiet -TargetName "fakeHost" -Count 1 | Should -BeFalse
```
```suggestion
            $result = $true
            { $result = Test-Connectiong -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" } | Should -Not -Throw
            $result | Should -BeFalse
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -59,18 +59,20 @@ Describe "Test-Connection" -tags "CI" {
 
-            { $result = Test-Connection "fakeHost" -Count 1 -Quiet -ErrorAction Stop } |
+        It "Ping fake host" {
+            { $result = Test-Connection "fakeHost" -Count 1 -ErrorAction Stop } |
```
```suggestion
            { Test-Connection "fakeHost" -Count 1 -ErrorAction Stop } |
```

vexx32 committed:
Apply suggestions from code review

Co-Authored-By: Ilya <darpa@yandex.ru>

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -59,18 +59,21 @@ Describe "Test-Connection" -tags "CI" {
-        It "Ping fake host" {
+        It 'returns false without errors for an unresolvable address when using -Quiet' {
+            $result = $true
+            { $result = Test-Connectiong -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" } | Should -Not -Throw
```
```suggestion
            { $result = Test-Connection -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" } | Should -Not -Throw
```

vexx32 replied:
Ha! Good catch. 😆

vexx32 committed:
Update test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1

Co-Authored-By: Ilya <darpa@yandex.ru>

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -732,6 +744,22 @@ private void ProcessPing(string targetNameOrAddress)
+        private IPHostEntry GetHostEntryWithCancel(string targetNameOrAddress)
+        {
+            var task = Dns.GetHostEntryAsync(targetNameOrAddress);
+            var waitHandles = new[] { ((IAsyncResult)task).AsyncWaitHandle, _dnsLookupCancel.Token.WaitHandle };
```
This works fine, but how about the following?
```c#
try
{
    var task = Dns.GetHostEntryAsync(targetNameOrAddress);
    task.Wait(_dnsLookupCancel.Token);
    return task.
}
catch (OperationCanceledException e)
{
    throw new PipelineStoppedException();
}
```
Of course, this will require the change of `catch (Exception ex)` to deal with `AggregateException` specially by getting the first item from the `InnerExceptions` collection

vexx32 replied:
Poked at this one too, the aggregate exception is kinda annoying to have to deal with... Hard to really say one is better than the other myself, I don't know enough, but it seems like they're roughly equivalent?

Saw a few SO posts mentioning that the `GetAwaiter().GetResult()` pattern can result in nicer stacktraces as well, so we might want to stick to that, possibly? 🤔

daxian-dbw replied:
`GetAwaiter().GetResult()` unwraps the aggregate exception for you, basically give you the first item of the `InnerExceptions` collection.
In our case here, the aggregate exception will only have one item in its `InnerExceptions` collection. It has more than 1 inner exceptions only when the task is an aggregation of multiple tasks, for example, `Task.WhenAll`.

vexx32 replied:
This seems to indicate it's a bit more than just unwrapping: https://stackoverflow.com/questions/36426937/what-is-the-difference-between-wait-vs-getawaiter-getresult/36427080

But yes, there shouldn't be a huge difference here in either case, from what I'm reading, though I've seen a few heated discussions arguing about using one v.s. the other. 😁

daxian-dbw replied:
> This seems to indicate it's a bit more than just unwrapping:

Good point. I missed that :)

vexx32 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -679,16 +682,25 @@ private void ProcessPing(string targetNameOrAddress)
                 }
+                catch (PipelineStoppedException p)
+                {
+                    throw p;
```
Didn't know that was a thing. Is there a significant difference?

daxian-dbw replied:
Which one you are asking about?

for the `throw`, it will keep the original stack trace; while `throw p` will rewrite the stack trace.
for the conditional catch block, it's slightly efficient because it avoids looking for a handler for the `PipelineStoppedException`. Although the efficiency doesn't matter here, it looks more concise :)

vexx32 replied:
Poked at this one for a bit and changed my mind a couple times, but it looks like the conditional catch block is about the same as just doing a general catch with an `if` statement when it comes down to it. Separate catch block seems a little tidier.

daxian-dbw replied:
I'm fine keeping the `catch (PipelineStoppedException p)` block, as long as you change `throw p` to `throw;`.
The main benefit of conditional exception is to avoid unnecessary search of exception handlers and the stack unwinding.

vexx32 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -732,6 +744,22 @@ private void ProcessPing(string targetNameOrAddress)
             return true;
         }
 
+        private IPHostEntry GetHostEntryWithCancel(string targetNameOrAddress)
```
Hey, we already have a `SendCancellablePing()` 😋 

No worries, I'll sort it out. 🙂

vexx32 committed:
:ok_hand: address review comments

vexx32 commented:
I think that should address all your comments @daxian-dbw. 🙂 

Let me know if I misunderstood anything; I'm still pretty new to fiddling with cancellation tokens and whatnot. ^^

vexx32 commented:
🤔 realizing that returning false from this method is still gonna be a bit janky. Need to handle it and actually write false values back in the other methods too... Hm. That's okay, shouldn't be too tricky to manage.

vexx32 committed:
:wrench: Make sure we're writing  quiet output

+ Since we're not emitting errors for these cases, we need to make sure
that we give the users an appropriate -Quiet response.

+ In most cases, this is just false. -MtuSize is a special case that typically
outputs integers, so write -1 to indicate failure to connect there.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -59,18 +59,21 @@ Describe "Test-Connection" -tags "CI" {
+        It 'returns false without errors for an unresolvable address when using -Quiet' {
+            $result = $true
+            { $result = Test-Connection -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" } | Should -Not -Throw
+            $result | Should -BeFalse
```
Since we expect result I suggest simplify the test:
```suggestion
            Test-Connection -Quiet -ErrorAction Stop -Count 1 -TargetName "fakeHost" | Should -BeFalse
```

vexx32 committed:
Update test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1

Co-Authored-By: Ilya <darpa@yandex.ru>

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:
```diff
@@ -58,6 +58,8 @@ public class TestConnectionCommand : PSCmdlet, IDisposable
 
         private static byte[]? s_DefaultSendBuffer;
 
+        private CancellationTokenSource _dnsLookupCancel = new CancellationTokenSource();
```
From Codacy:
```suggestion
        private readonly CancellationTokenSource _dnsLookupCancel = new CancellationTokenSource();
```

vexx32 replied:
Huh, I thought I... ah, I must have taken it out again at some point. Thanks!

vexx32 committed:
Update src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

Co-Authored-By: Ilya <darpa@yandex.ru>
